### DISCUSSION TITLE
Documentation for RPC Protocol

### DISCUSSION THREAD

**Discussion by ecyht2** [ggml-org/llama.cpp#9446] [General]
Is there any documentation for the RPC protocol? I think it would be nice to have it documented. I have an idea on how it works by reading the source code but I think it would be helpful to have a formal documentation about it. I am thinking of implementing my own server just for fun.

**Comment by rgerganov** [comment]
The RPC protocol is quite simple. There are several [RPC commands](https://github.com/ggerganov/llama.cpp/blob/master/ggml/src/ggml-rpc.cpp#L85-L96) which are used to implement the backend interfaces defined in [ggml-backend-impl.h](https://github.com/ggerganov/llama.cpp/blob/master/ggml/src/ggml-backend-impl.h).

The [format](https://github.com/ggerganov/llama.cpp/blob/master/ggml/src/ggml-rpc.cpp#L253-L254) of the RPC request and RPC response is:
```
RPC request:
 | rpc_cmd (1 byte) | request_size (8 bytes) | request_data (request_size bytes) |

RPC response:
 | response_size (8 bytes) | response_data (response_size bytes) |
```

The RPC server is handling one client at a time, in a [single thread](https://github.com/ggerganov/llama.cpp/blob/master/ggml/src/ggml-rpc.cpp#L1213-L1224). There is some [work-in-progress](https://github.com/ggerganov/llama.cpp/pull/8032) for server-to-server communication but it is not ready yet.

That's it. If you have any questions, I'd be happy to answer.
